pkg/service/keystore/storage.go

```diff
@@ -48,43 +56,110 @@ const (
 )
 
 type Storage struct {
-	db         storage.ServiceStorage
-	serviceKey []byte
+	db        storage.ServiceStorage
+	encrypter Encrypter
+	decrypter Decrypter
+}
+
+func NewKeyStoreStorage(db storage.ServiceStorage, e Encrypter, d Decrypter) (*Storage, error) {
+	s := &Storage{
+		db:        db,
+		encrypter: e,
+		decrypter: d,
+	}
+
+	return s, nil
+}
+
+type wrappedEncrypter struct {
+	tink.AEAD
 }
 
-func NewKeyStoreStorage(db storage.ServiceStorage, key ServiceKey) (*Storage, error) {
-	keyBytes, err := base58.Decode(key.Base58Key)
+func (w wrappedEncrypter) Encrypt(_ context.Context, plaintext, contextData []byte) ([]byte, error) {
+	return w.AEAD.Encrypt(plaintext, contextData)
+}
+
+type wrappedDecrypter struct {
+	tink.AEAD
+}
+
+func (w wrappedDecrypter) Decrypt(_ context.Context, ciphertext, contextInfo []byte) ([]byte, error) {
+	return w.AEAD.Decrypt(ciphertext, contextInfo)
+}
+
+const (
+	gcpKMSScheme = "gcp-kms"
+	awsKMSScheme = "aws-kms"
+)
+
+func NewEncryption(db storage.ServiceStorage, cfg config.KeyStoreServiceConfig) (Encrypter, Decrypter, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if len(cfg.MasterKeyURI) != 0 {
+		return NewExternalEncrypter(ctx, cfg)
+	}
+
+	// First, generate a service key
+	serviceKey, serviceKeySalt, err := GenerateServiceKey(cfg.MasterKeyPassword)
 	if err != nil {
-		return nil, errors.Wrap(err, "could not decode service key")
+		return nil, nil, sdkutil.LoggingErrorMsg(err, "generating service key")
+	}
+
+	key := ServiceKey{
+		Base58Key:  serviceKey,
+		Base58Salt: serviceKeySalt,
+	}
+	if err := storeServiceKey(ctx, db, key); err != nil {
+		return nil, nil, err
 	}
-	bolt := &Storage{db: db, serviceKey: keyBytes}
+	return &encrypter{db}, &decrypter{db}, nil
+}
 
-	// first, store the service key
-	if err = bolt.storeServiceKey(context.Background(), key); err != nil {
-		return nil, errors.Wrap(err, "could not store service key")
+func NewExternalEncrypter(ctx context.Context, cfg config.KeyStoreServiceConfig) (Encrypter, Decrypter, error) {
+	var client registry.KMSClient
+	var err error
+	switch {
+	case strings.HasPrefix(cfg.MasterKeyURI, gcpKMSScheme):
+		client, err = gcpkms.NewClientWithOptions(ctx, cfg.MasterKeyURI, option.WithCredentialsFile(cfg.KMSCredentialsPath))
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "creating gcp kms client")
+		}
+	case strings.HasPrefix(cfg.MasterKeyURI, awsKMSScheme):
+		client, err = awskms.NewClientWithCredentials(cfg.MasterKeyURI, cfg.KMSCredentialsPath)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "creating aws kms client")
+		}
+	default:
+		return nil, nil, errors.Errorf("master_key_uri value %q is not supported", cfg.MasterKeyURI)
+	}
+	registry.RegisterKMSClient(client)
+	dek := aead.XChaCha20Poly1305KeyTemplate()
```

Comment:
so we are assuming they have a XChaCha20Poly1305 key, if we're forcing this (I think we should for now at least) then this needs to be documented on how they configure their KMS